Dart/src/com/jetbrains/lang/dart/analyzer/DartAnalysisServerService.java

```diff
@@ -1481,6 +1486,53 @@ public void onError(@NotNull final RequestError error) {
     return resultRef.get();
   }
 
+  @Nullable
+  public CompletionInfo2 completion_getSuggestions2(@NotNull final VirtualFile file, final int _offset, final int maxResults) {
+    final AnalysisServer server = myServer;
+    if (server == null) {
+      return null;
+    }
+
+    for (DartCompletionTimerExtension extension : DartCompletionTimerExtension.getExtensions()) {
+      extension.dartCompletionStart();
+    }
+
+    final String filePath = FileUtil.toSystemDependentName(file.getPath());
+    final Ref<CompletionInfo2> resultRef = new Ref<>();
+    final CountDownLatch latch = new CountDownLatch(1);
+    final int offset = getOriginalOffset(file, _offset);
+
+    server.completion_getSuggestions2(filePath, offset, maxResults, new GetSuggestionsConsumer2() {
+      @Override
+      public void computedSuggestions(int replacementOffset,
+                                      int replacementLength,
+                                      List<CompletionSuggestion> suggestions,
+                                      List<String> libraryUrisToImport,
+                                      boolean isIncomplete) {
+        resultRef.set(new CompletionInfo2(replacementOffset, replacementLength, suggestions, libraryUrisToImport, isIncomplete));
+        latch.countDown();
+      }
+
+      @Override
+      public void onError(@NotNull final RequestError error) {
+        for (DartCompletionTimerExtension extension : DartCompletionTimerExtension.getExtensions()) {
+          extension.dartCompletionError(StringUtil.notNullize(error.getCode()), StringUtil.notNullize(error.getMessage()),
+                                        StringUtil.notNullize(error.getStackTrace()));
+        }
+        // Not a problem. Happens if a file is outside the project, or server is just not ready yet.
+        latch.countDown();
+      }
+    });
+
+    awaitForLatchCheckingCanceled(server, latch, GET_SUGGESTIONS_TIMEOUT);
```

Comment:
I guess we can use a timeout longer than 5 seconds for this case. This is a ProcessCancelledException-protected code that runs in a background thread so we can afford a longer 'in progress' spinner in code completion popup if server might need more time to calculate and send results.